### ISSUE TITLE
Minimum stake adjustment

### ISSUE THREAD

**Issue by raphaelstar** [freqtrade/freqtrade#8093]
## Describe your environment

  * Operating system: linux
  * Python Version: 2.10
  * CCXT version: 2.5.39
  * Freqtrade Version: 2023.1
  
## Your question

I am trying to understand the logic behind the `amount_reserve_percent` here: https://github.com/freqtrade/freqtrade/blob/ceaaac6c3aebd6a7afd852683e26b0ad167da90f/freqtrade/exchange/exchange.py#L788

I have read your previous comment here: #5856

> If you got to the website and try to trade it, you'll notice that the minimum tradable size is 10 KRL.
10 KRL will result in ~40$ with the price your bot tried to buy.
Freqtrade must however make sure that there's still some room below the trade to place a stoploss (otherwise you'll have 100% loss if it drops below tradable size).

I don't understand this logic. If one open a trade with the minimum amount of 10 KRL, with a stoploss of 10%. If KRL losses 10%, the trade can still be closed anyway by selling those 10 KRL back, the stoploss is triggered and doesn't results in a loss of 100%. There is no need to buy more than 10 KRL to "have some room below"?

**Comment by xmatthias** [comment]
Exchanges usually have 2 limits - one on the amount - and one on value (there's actually a 3rd limit on price, but that's not relevant here).

So the limit is either a minimum of 10 KRL - or a minimum value of 10$ (for example).
while a 10% drop will still have you own 10 KRL - it might be below the 10$ value - which means you can't sell it anymore.

you're also ignoring fees, which different exchanges handle differently (you might buy 10 KRL on binance - but if you don't own BNB - you'll only get 9.9 KRL, the 0.1 KRL is fees.
Now the worst case in that case is if KRL trades in lot-sizes of 1 - which means you can only sell 9 KRL, and the 0.9 KRL remain on the account (also called binance dust).

While the stoploss will protect you from this (to some degree) - a stoploss is not guaranteed to execute EXACTLY at the value it's placed (depending on the order type, it could be several % down - if it's an illiquid market with market orders.
 
As such - an offset of 5% is applied - meaning you cannot go to the very bare minimum of the tradable amounts.
Users can configure this - but I'd recommend not to.

**Comment by raphaelstar** [comment]
Ok, I understand, you need to account for the 2 limits: limit in amount, and limit in cost. But it seems that the stoploss reserve should be added only to the cost limit:

Let's say the amount limit is 10 KRL, and the cost limit is 10$.
With:
Stoploss: 10%
Price: KRL/USD = 10$

The minimal stake amount should be 100$ (cost for 10 KRL) because even if there is a drop 10%, you will still be able to sell the 10 KRL for 90$.

Currently, the stoploss margin is applied in on both limits and is resulting in an unnecessary higher minimum stake that would be 111$.

Here is my suggestion for the stake limit calculation:
```python

    def _get_stake_amount_limit(
        self,
        pair: str,
        price: float,
        stoploss: float,
        limit: Literal['min', 'max'],
        leverage: Optional[float] = 1.0
    ) -> Optional[float]:

        isMin = limit == 'min'

        try:
            market = self.markets[pair]
        except KeyError:
            raise ValueError(f"Can't get market information for symbol {pair}")
        
        if isMin:
          margin_reserve = 1 + self._config.get('amount_reserve_percent',
                           DEFAULT_AMOUNT_RESERVE_PERCENT)
          # Cap the stop loss reserve to 50%
          stoploss = min(abs(stoploss), 0.5)
          stoploss_reserve = 1 / (1 - stoploss)
        else:
          margin_reserve = 1.0
          stoploss_reserve = 1.0
          
        stake_limits = []
        limits = market['limits']
        if (limits['cost'][limit] is not None):
            stake_limits.append(
                self._contracts_to_amount(
                    pair,
                    limits['cost'][limit]
                ) * margin_reserve * stoploss_reserve
            )

        if (limits['amount'][limit] is not None):
            stake_limits.append(
                self._contracts_to_amount(
                    pair,
                    limits['amount'][limit] * price * margin_reserve
                )
            )

        if not stake_limits:
            return None if isMin else float('inf')

        # The value returned should satisfy both limits: for amount (base currency) and
        # for cost (quote, stake currency), so max() is used here.
        # See also #2575 at github.
        return self._get_stake_amount_considering_leverage(
            max(stake_limits),
            leverage or 1.0
        ) if isMin else min(stake_limits)
```

**Comment by xmatthias** [comment]
You're ignoring fees paid in receiving currency.
I don't see any need for change in this regard - and i don't see why it should be a problem either.

Please note: If you write a change - that should be a PR (which also explains WHY the change should be necessary, what problems it aims to fix) - and allows a direct comparison on what actually changed.
I'll not copy/paste it to see how it differs from the existing function. That's work i don't see why i should do - for something i don't see a need to change either.

A guide for the github PR workflow can be found [here](https://guides.github.com/introduction/flow/).

**Comment by raphaelstar** [comment]
> I'll not copy/paste it to see how it differs from the existing function. That's work i don't see why i should do - for something i don't see a need to change either.

No problem, ignore my code suggestion. It was meant to help, not to burden you.

> You're ignoring fees paid in receiving currency.

Yes, I ignored the fees in my example voluntarily for simplification. 

> I don't see any need for change in this regard - and i don't see why it should be a problem either.

I don't get your point, you don't see any need for optimizing the minimum stake value?

**Comment by xmatthias** [comment]
you can set the offset to 0 if you'd like to go without offset.
the offset is meant to ensure the position remains sellable. 
It's a protective measure - optimizing for the very last cent is risky - as then rounding can impact this as well.

Usual position limits are around 10-15$ - optimizing this value by 5% while at the same time, risking to lose it all (you can't sell it anymore) is a risk that doesn't make sense in my eyes.

As such - i don't see why this offset / intentional inaccuracy is causing a problem - and therefore also don't see a need for change.


Maybe you can describe the problem you're facing due to this - but for now - i don't see this as problematic.

**Comment by raphaelstar** [comment]
I think you didn't understand my point. I am not arguing against the 5% `amount_reserve_percent` margin. But against the `stoploss` margin that is applied in every cases. 

Let my try to reformulate my example:

Let's say the amount limit is 10 KRL, and the cost limit is 10$.
With:
Stoploss: 10%
Price: KRL/USD = 10$

Currently the minimum stake would be 111$ (+5% reserve)
What I try to show is this can be reduced to only 100$ (+5% reserve)

Because:
If the buy amount is 100$ (10 KRL) and the stoploss is triggered: there is a 10% drop in value. It is still possible to sell the 10 KRL for 90$ because it is still above the cost limit of 10$.

The potential optimization is not in the order of (5%) but in the order of stoploss% which can be significant.

**Comment by raphaelstar** [comment]
> Maybe you can describe the problem you're facing due to this - but for now - i don't see this as problematic.

Indeed, I am facing a problem. Since I am using a stoploss of 30%, the minimum stake amount overshoot and prevents entering trades:
`Adjusted stake amount for pair LTC/USDT:USDT is more than 30% bigger than the desired stake amount of (100.00000000 * 1.3 = 130.00000000) < 132.16875000000002), ignoring trade.`
